lib/adversarial_da.py

```diff
@@ -0,0 +1,187 @@
+"""Discriminative domain adaptation module.
+"""
+import tensorflow as tf
+import tensorflow.keras as tkf
+
+def get_discriminator(input_shape, num_layers=3, activation="relu", dropout=0):
+  """Returns the disctiminator model.
+
+  Arguments:
+  input_shape: shape of the input tensor.
+  num_layers: number of convolutional layers.
+  activaiton: activation function (default is relu).
+  """
+  discriminator = tkf.Sequential(name="discriminator")
+  discriminator.add(tkf.layers.InputLayer(input_shape))
+  for x in range(num_layers):
+    discriminator.add(
+        tkf.layers.Conv2D(16 * (x + 1), (3, 3), strides=(1, 1),
+                          padding="same", activation=activation))
+
+  discriminator.add(tkf.layers.AveragePooling2D(pool_size=(8, 8)))
+  discriminator.add(tkf.layers.Flatten())
+  discriminator.add(tkf.layers.Dropout(dropout))
+  discriminator.add(tkf.layers.Dense(1, activation="sigmoid"))
+  return discriminator
+
+class AdversarialClassifier(tkf.Model):
+  """Discriminative classifier for unsupervised domain adaptation.
+
+  Consists of a classifier network that is trained to perform classification on
+  the source domain, and domain disriminator that, given late features from the
+  classifier, predicts whether the features come from source or target domain.
+  The target domain mixture weights are optimized to fool the discriminator.
+
+  Arguments:
+  classifier: classifier model.
+  discriminator: discriminator model.
+  domains: list of domain names.
+  target_domain: target domain name.
+  num_layers: number of resblock layers in the classifier.
+  """
+  def __init__(self, classifier, discriminator, domains, target_domain,
+               num_layers, **args):
+    super(AdversarialClassifier, self).__init__(**args)
+    self.classifier = classifier
+    self.discriminator = discriminator
+    self.domains = domains
+    self.num_domains = len(self.domains)
+    self.target_domain = target_domain
+    if target_domain in domains:
+      self.target_domain_idx = int(self.domains.index(self.target_domain))
+    else:
+      raise ValueError("The target domain is not in the domains list.")
+    self.num_layers = num_layers
+    self._input_shape = self.classifier.input_shape[0][1:]
+
+    self._set_feature_extractor()
+
+
+  def _set_feature_extractor(self):
+    """Sets late feature extractor of the model."""
+    cls_inputs = [tkf.Input(name="%s_in_new" % x, shape=self._input_shape)
+                  for x in self.domains]
+    new_model_out = self.classifier(cls_inputs)
+    try:
+      block = (self.classifier.get_layer("feature_extractor")
+                              .get_layer("weighted_resblock"))
+      features_output_arr = []
+      for idx in range(self.num_domains):
+        out_idx = self.num_layers * (2 * self.num_domains - idx + 1) - 1
+        outputs = block.get_output_at(out_idx)
+        features_output_arr.append(outputs)
+    except ValueError:
+      #finding the last separate batch norm block
```

Comment:
One advice I was given regarding comment like these: generally prefer comments that answer the question "why?". If the comment answers the question "what?", it might be an indication that you need to create a self-contained function, or variable with an understandable name (code comments itself sort of argument). Just an advice, don't have to take it into account here.